### PR TITLE
Fix overflow issues when unpacking large numbers

### LINKED ISSUE #33111: torch.tensor([0.1, 999999999999999999999]) fails with "Overflow when unpacking double"
Work around is to add a decimal to the end of the big number.

### PR DESCRIPTION
**Pull request by RockingJavaBean**
Resolve #33111

relax the overflow and precision lost checks when unpacking doubles.

Signed-off-by: Xiong Wei <xiongw.fnst@cn.fujitsu.com>

### CODE REVIEW

**Comment by RockingJavaBean** [comment]
Kindly request for review. @ezyang

**Comment by ezyang** [comment]
Looking over the PR here, I'm not sure your fix is complete (and it also reveals that, in a way, the old torch behavior is consistent, though not consistent with Numpy). 

Your new code continues to do the "precision loss" check, and you simply omitted the overflow check. But this seems really weird: this means that there is a nether region where you haven't overflowed, but you are too large to accurately represent in float, and Torch will throw an error (but if you made it smaller or bigger, no error!)

Although torch's old behavior is "consistent", Numpy doesn't seem to do an overflow test at all:

```
>>> numpy.array([1., 9999999999999999999999999], dtype='float')
array([  1.00000000e+00,   1.00000000e+25])
```

So it might still be a good idea from a Numpy compatibility standpoint to relax the check here. But I'll defer to @mruberry on this matter.

**Review by ezyang — CHANGES_REQUESTED** [review]
Don't think this is complete

**Comment by RockingJavaBean** [comment]
Thank you so much for pointing out the problems of the previous PR.

I have removed both the overflow and "precision lost" checks to prevent such nether regions.

Meanwhile, I found `THPUtils_unpackDouble` does not check the exceptions of `PyLong_AsLongLongAndOverflow` as other unpacking functions, such as `THPUtils_unpackLong`.

The corresponding checking code is also added to the current PR. 
Please help evaluate whether it is feasible to relax the checks when unpacking doubles.

@ezyang @mruberry

**Comment by ezyang** [comment]
Still waiting on @mruberry to chime in. While we wait, do you mind testing for the specific cases where the int is not exactly representable as double?

File: test/test_torch.py [hunk 1]
```diff
@@ -1223,6 +1223,11 @@ def test_new_tensor(self):
             self.assertEqual(res2, expected, exact_dtype=False)
             self.assertIs(torch.int, res2.dtype)
 
+            # test unpack large number
+            res3 = torch.tensor([0.1, 999999999999999999999])
```
Comment (mruberry) [inline test/test_torch.py:1227]: Let's do something like this:

```
if TEST_NUMPY:
  vals = (2**24 + 1, 2**53 + 1,  np.iinfo(np.int64).max, np.iinfo(np.uint64).max, np.iinfo(np.uint64).max + 1, -1e500, 
          1e500)
  dtypes = (torch.float32, torch.float64)
  for val, dtype in product(vals, dtypes):
    t = torch.tensor(val, dtype=dtype)
    a = np.array(val, dtype=torch_to_numpy_dtype_dict[dtype])
    self.assertEqual(t, torch.from_numpy(a))
```

Ideally this test would be refactored to be device generic so we could test it on CPU and CUDA easily, too.

File: torch/csrc/utils/python_numbers.h [hunk 2]
```diff
@@ -109,13 +109,12 @@ inline double THPUtils_unpackDouble(PyObject* obj) {
   if (PyLong_Check(obj)) {
     int overflow;
     long long value = PyLong_AsLongLongAndOverflow(obj, &overflow);
```
Comment (mruberry) [inline torch/csrc/utils/python_numbers.h:111]: I'm not sure this part of the function is needed anymore. What would happen if we immediately converted using PyFloat_AsDouble?

**Review by mruberry — CHANGES_REQUESTED** [review]

**Comment by mruberry** [comment]
> Still waiting on @mruberry to chime in. While we wait, do you mind testing for the specific cases where the int is not exactly representable as double?

Sorry, was swamped yesterday. I suggested an expanded test.

File: test/test_torch.py (hunk 1 above)
Comment (RockingJavaBean) [reply test/test_torch.py:1227]: Thanks for the tip, I will implement a device generic test method based on the provided code snippet.

File: torch/csrc/utils/python_numbers.h (hunk 2 above)
Comment (RockingJavaBean) [reply torch/csrc/utils/python_numbers.h:111]: I tested removing the `PyLong_Check` part of code and immediately converting using `PyFloat_AsDouble` locally, it has the same result as casting the result of `PyLong_AsLongLongAndOverflow` to `double` when overflow and precision lost checks pass.

**Review by mruberry — CHANGES_REQUESTED** [review]
My mistake as I wasn't clear enough in my last message.

Instead of this approach, don't update common_utils.py and put your test in TestTorchDeviceType: 

https://github.com/pytorch/pytorch/blob/5cfd1a190e19b30478f5e584052261dd147a08e9/test/test_torch.py#L5385

We have an existing mechanism to let your test run on all available devices (see the Wiki article https://github.com/pytorch/pytorch/wiki/Writing-tests-that-run-on-all-available-device-types). Your test signature will be:

```
@unittest.skipIf(not TEST_NUMPY, "NumPy not found")
@dtypes(torch.float32, torch.float64)
def test_unpack_double(self, device, dtype):
```

And you won't need to check TEST_NUMPY or enumerate the dtypes yourself anymore. You also won't need a helper function as the test suite will instantiate all the variants of your test for you. Let me know if you have any questions.

**Comment by RockingJavaBean** [comment]
@mruberry thanks for explaining how to write device generic tests in detail.
I have updated the test method for unpacking doubles accordingly.

**Review by mruberry — APPROVED** [review]
Thanks @RockingJavaBean, nice work! Thanks for taking the time to learn about the test framework.

**Comment by mruberry** [comment]
@ezyang Would you like to take a look?

**Review by ezyang — APPROVED** [review]
